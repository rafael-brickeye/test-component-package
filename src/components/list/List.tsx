import React from 'react';
import {ActivityIndicator, FlatList, FlatListProps, ViewStyle} from 'react-native';
import useStyles from './List.styles';
import {color} from '../../themes';

// TODO:
// will need to revisit and update the interface and props of this
// component when pagination is implemented for the Events List API
// and for future API refactors to support data paginations

export type ListPropTypes = {
  isLoading: boolean;
  footerTestId?: string;
} & FlatListProps<any> &
  ViewStyle;

const List: React.FC<ListPropTypes> = ({isLoading, footerTestId, style: styleProps, ...props}) => {
  const style = useStyles({...props});

  return (
    <FlatList
      ListFooterComponent={() => <LoaderComponent footerTestId={footerTestId} isLoading={isLoading} />}
      style={{...style.listContainer, ...(styleProps as ViewStyle)}}
      {...props}
    />
  );
};

const LoaderComponent = ({isLoading, footerTestId}: {footerTestId?: string; isLoading: boolean}) => {
  const {
    brand: {
      protect: {primary},
    },
  } = color;

  if (!isLoading) return null;

  return <ActivityIndicator testID={footerTestId} size={'large'} color={primary.main} />;
};

export {List};
