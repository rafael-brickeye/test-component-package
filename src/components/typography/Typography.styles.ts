import {StyleSheet} from 'react-native';
import {getDefaultStylesFromProps, sizing, color as themeColor} from '../../themes';

const useStyles = ({...props}) => {
  const styles = getDefaultStylesFromProps(props);

  return StyleSheet.create({
    typographyContainer: {
      fontSize: sizing.font.md,
      fontWeight: 'normal',
      color: themeColor.text.black,
      ...styles,
    },
  });
};

export default useStyles;
