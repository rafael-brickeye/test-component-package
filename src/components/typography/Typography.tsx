import React from 'react';
import {Text, TextProps, TextStyle} from 'react-native';
import useStyles from './Typography.styles';

type TextDefaultProps = TextProps & TextStyle;

export type TypographyPropTypes = {
  truncateEnd?: boolean;
} & TextDefaultProps;

const Typography: React.FC<TypographyPropTypes> = ({children, truncateEnd = false, style: styleProps, ...props}) => {
  const style = useStyles({...props});

  return (
    <Text
      numberOfLines={truncateEnd ? 1 : 0}
      style={{...style.typographyContainer, ...(styleProps as TextStyle)}}
      {...props}>
      {children}
    </Text>
  );
};

export {Typography};
