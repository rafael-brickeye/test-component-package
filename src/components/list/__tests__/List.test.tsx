import React from 'react';
import {cleanup, render} from '@testing-library/react-native';
import '@testing-library/jest-native';
import {List, ListPropTypes} from '../List';
import {Typography} from '../../typography';

afterEach(cleanup);

const renderItem = ({item}: {item: {id: number; title: string; testId: string}}) => (
  <Typography testID={item.testId}>{item.title}</Typography>
);
const keyExtractor = (item: {id: number}) => item.id.toString();

const defaultParams = {
  data: [
    {id: 1, title: 'hello', testId: 'test-component-item-1'},
    {id: 2, title: 'world', testId: 'test-component-item-2'},
  ],
  footerTestId: 'footer-component',
  isLoading: false,
  keyExtractor,
  renderItem,
};

const ListComponent = (props: Partial<ListPropTypes>) => <List {...defaultParams} {...props} />;

describe('Base List Component', () => {
  it('should match the existing snapshot', () => {
    const snapshot = render(<ListComponent />).toJSON();

    expect(snapshot).toMatchSnapshot();
  });

  it('should be able to see the list of items passed from renderItem', () => {
    const {getByTestId} = render(<ListComponent />);

    expect(getByTestId('test-component-item-1')).toBeVisible();
    expect(getByTestId('test-component-item-1')).toHaveTextContent('hello');
    expect(getByTestId('test-component-item-2')).toBeVisible();
    expect(getByTestId('test-component-item-2')).toHaveTextContent('world');
  });

  it('should be able to render Footer Loader component when List component has isLoading set to TRUE', () => {
    const {getByTestId} = render(<ListComponent isLoading={true} />);

    expect(getByTestId(defaultParams.footerTestId)).toBeVisible();
  });
});
