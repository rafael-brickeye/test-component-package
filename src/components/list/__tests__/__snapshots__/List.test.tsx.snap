// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base List Component should match the existing snapshot 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "id": 1,
        "testId": "test-component-item-1",
        "title": "hello",
      },
      Object {
        "id": 2,
        "testId": "test-component-item-2",
        "title": "world",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={Object {}}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        numberOfLines={0}
        style={
          Object {
            "color": "#000000DE",
            "fontSize": 12,
            "fontWeight": "normal",
          }
        }
        testID="test-component-item-1"
      >
        hello
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        numberOfLines={0}
        style={
          Object {
            "color": "#000000DE",
            "fontSize": 12,
            "fontWeight": "normal",
          }
        }
        testID="test-component-item-2"
      >
        world
      </Text>
    </View>
    <View
      onLayout={[Function]}
    />
  </View>
</RCTScrollView>
`;
